source/contributing/implementation/datamanipulator.rst

```diff
@@ -0,0 +1,467 @@
+whether the operation was successful or not.
+
+.. code-block:: java
+
+    protected boolean set(EntityLivingBase entity, Map<Key<?>, Object> keyValues) {
+        entity.getEntityAttribute(SharedMonsterAttributes.maxHealth)
+            .setBaseValue(((Double) keyValues.get(Keys.MAX_HEALTH)).floatValue());
+        entity.setHealth(((Double) keyValues.get(Keys.HEALTH)).floatValue());
+        return true;
+    }
+
+.. tip::
+
+    To understand ``DataTransactionResult`` \ s, check the :doc:`corresponding docs page
+    <../../plugin/data/transactions>` and refer to the `DataTransactionBuilder API
+    Docs <https://jd.spongepowered.org/index.html?org/spongepowered/api/data/DataTransactionBuilder.html>`_ to
+    create one.
+
+Removal Methods
+~~~~~~~~~~~~~~~
+
+The ``remove()`` method attempts to remove data from the ``DataHolder`` and returns a ``DataTransactionResult``.
+
+Since it is impossible for an ``EntityLivingBase`` to not have any health, this operation will always fail on
+the ``HealthDataProcessor``.
+
+.. code-block:: java
+
+    public DataTransactionResult remove(DataHolder dataHolder) {
+        return DataTransactionBuilder.failNoData();
+    }
+
+
+Getter Methods
+~~~~~~~~~~~~~~
+
+Getter methods obtain data from a ``DataHolder`` and return an optional ``DataManipulator``. The
+``DataProcessor`` interface specifies the methods ``from()`` and ``createFrom()``, the difference being that
+``from()`` will return ``Optional.empty()`` if the data holder is compatible, but currently does not contain the
+data, while ``createFrom()`` will provide a ``DataManipulator`` holding default values in that case.
+
+Again, ``AbstractEntityDataProcessor`` will provide most of the implementation for this and only requires a
+method to get the actual values present on the ``DataHolder``. This method is only called after ``supports()``
+and ``doesDataExist()`` both returned true.
+
+.. code-block:: java
+
+    protected Map<Key<?>, ?> getValues(EntityLivingBase entity) {
+        final double health = entity.getHealth();
+        final double maxHealth = entity.getMaxHealth();
+        return ImmutableMap.<Key<?>, Object>of(Keys.HEALTH, health, Keys.MAX_HEALTH, maxHealth);
+    }
+
+Filler Methods
+~~~~~~~~~~~~~~
+
+A filler method is different from a getter method in that it receives a ``DataManipulator`` to fill with values.
+These values either come from a ``DataHolder`` or have to be deserialized from a ``DataContainer``. The method
+returns ``Optional.empty()`` if the ``DataHolder`` is incompatible.
+
+``AbstractEntityDataProcessor`` already handles filling from ``DataHolders`` by creating a ``DataManipulator``
+from the holder and then merging it with the supplied manipulator, but the ``DataContainer`` deserialization it
+can not provide.
+
+.. code-block:: java
+
+    public Optional<HealthData> fill(DataContainer container, HealthData healthData) {
+        healthData.set(Keys.MAX_HEALTH, DataUtil.getData(container, Keys.MAX_HEALTH));
```

Comment:
change this such that it checks if the `DataContainer` actually contains the keys, if it does, then just get the data directly. otherwise this is throwing `InvalidDataException`s.